packages/eslint-plugin-prairielearn/src/rules/safe-db-types.ts

```diff
@@ -0,0 +1,592 @@
+                        node: attribute || reportNode,
+                        messageId: 'unsafeTypes',
+                        data: { propName, typeName },
+                      });
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Check for spread attributes
+  const attributes = childOpeningElement.attributes;
+  for (const attr of attributes) {
+    if (attr.type === 'JSXSpreadAttribute') {
+      context.report({
+        node: reportNode,
+        messageId: 'spreadAttributes',
+      });
+      continue;
+    }
+  }
+}
+
+export default ESLintUtils.RuleCreator.withoutDocs<
+  [{ allowDbTypes?: (string | RegExp)[] }?],
+  'spreadAttributes' | 'unsafeTypes'
+>({
+  meta: {
+    type: 'problem',
+    messages: {
+      spreadAttributes: 'Spread attributes are not allowed in Hydrate children.',
+      unsafeTypes:
+        'Prop "{{propName}}" uses type "{{typeName}}" which is derived from db-types.ts. Use safe-db-types.ts instead.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allowDbTypes: {
+            type: 'array',
+            items: {
+              type: 'string',
+            },
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [{}],
+  create(context) {
+    const options = context.options[0] || {};
+    const allowlist = options.allowDbTypes || [];
+    return {
+      JSXElement(node) {
+        const openingElementNameExpression = node.openingElement.name;
+
+        // TODO: handle namespace and member expressions?
+        if (openingElementNameExpression.type !== 'JSXIdentifier') return;
+
+        const elementName = openingElementNameExpression.name;
+        if (elementName !== 'Hydrate') return;
+
+        const child = extractChild(node.children);
+        if (!child) return;
+
+        // Get the component being rendered
+        const childOpeningElement = child.openingElement;
+        const childElementName = childOpeningElement.name;
+
+        if (childElementName.type !== 'JSXIdentifier') return;
+
+        // Get the component's type to inspect its props
+        const services = ESLintUtils.getParserServices(context);
+        const typeChecker = services.program.getTypeChecker();
+        const tsChildNode = services.esTreeNodeToTSNodeMap.get(childElementName);
+        const componentSymbol = typeChecker.getSymbolAtLocation(tsChildNode);
+
+        if (!componentSymbol) return;
+
+        checkComponentProps(
+          context,
+          typeChecker,
+          componentSymbol,
+          tsChildNode,
+          child,
+          child,
+          allowlist,
+        );
+      },
+
+      CallExpression(node) {
+        // Check for hydrateHtml(<Component ... />, props?) calls
+        if (node.callee.type !== 'Identifier' || node.callee.name !== 'hydrateHtml') return;
```

Comment:
I don't like burying details of our own setup in the lint rule. If we have custom fields/things to check, can we extract those to the top of the lint rule as constants?